ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+    // that might have been queued previously before adding this new one.
+    flush_queue(heap_recorder);
+  } else {
+    // We weren't able to get a lock, so enqueue this sample for later processing
+    // and end early
+    if (error == EBUSY) {
+      enqueue_allocation(heap_recorder, heap_stack, new_obj, weight);
+    } else {
+      ENFORCE_SUCCESS_GVL(error)
+    }
+    return;
+  }
+
+  // If we got this far, we got a write lock so we can commit the record
+  commit_allocation(heap_recorder, heap_stack, new_obj, weight);
+
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+void record_heap_free(heap_recorder *heap_recorder, VALUE obj) {
+  object_record *object_record = NULL;
+  st_lookup(heap_recorder->object_records, (st_data_t) obj, (st_data_t*) &object_record);
+
+  if (object_record == NULL) {
+    // we don't seem to be tracking this object on the table atm
+    // check if the allocation sample is in the queue
+    for (size_t i = 0; i < heap_recorder->queued_samples_len; i++) {
+      sample *queued_sample = &heap_recorder->queued_samples[i];
+      if (queued_sample->obj == obj) {
+        queued_sample->skip = true;
+        break;
+      }
+    }
+
+    // free of an untracked object, return early
+    return;
+  }
+
+  // if we got this far, we freed a tracked object so need to update records!
+  int error = pthread_mutex_trylock(&heap_recorder->records_mutex);
+  if (error) {
+    // We weren't able to get a lock, so enqueue this sample for later processing
+    // and exit early
+    if (error == EBUSY) {
+      enqueue_free(heap_recorder, obj);
+    } else {
+      ENFORCE_SUCCESS_GVL(error)
+    }
+    return;
+  }
+
+  // If we got this far, we got a write lock so we can commit the record
+  commit_free(heap_recorder, obj);
+
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+// ===============
+// Heap Record API
+// ===============
+heap_record* heap_record_init(const heap_stack *stack) {
+  heap_record* record = ruby_xcalloc(1, sizeof(heap_record));
+  record->num_tracked_objects = 0;
+  record->inuse_objects = 0;
+  record->stack = stack;
+  return record;
+}
+
+void heap_record_free(heap_record *record) {
+  ruby_xfree(record);
+}
+
+
+// =================
+// Object Record API
+// =================
+object_record* object_record_init(VALUE new_obj, unsigned int weight, heap_record *heap_record) {
+  object_record* record = ruby_xcalloc(1, sizeof(object_record));
+  record->weight = weight;
+  record->heap_record = heap_record;
+  return record;
+}
+
+void object_record_free(object_record *record) {
+  ruby_xfree(record);
+}
+
+// ==============
+// Heap Frame API
+// ==============
+int heap_frame_cmp(heap_frame *f1, heap_frame *f2) {
+  int cmp = strcmp(f1->name, f2->name);
+  if (cmp != 0) {
+    return cmp;
+  }
+  cmp = strcmp(f1->filename, f2->filename);
+  if (cmp != 0) {
+    return cmp;
+  }
+  return (int) (f1->line - f2->line);
+}
```

Comment:
Minor: May be worth starting with the line first, since it's the cheapest check.